doc: document support for package.json fields (#34970)
nodejs/node: doc/api/packages.md

until the next folder containing another `package.json` file, are a _package
scope_.

> ## Determining module system

Reviewer: I think “Determining module system” needs to come first, above any other section. The most important thing to know as a package author, even before the `package.json` fields, is that there are two module systems; and then when we introduce `"type"` and `"exports"` and those sections reference CommonJS and ESM, the reader already knows what those systems are.

> ## `package.json` supported fields

Reviewer: This should maybe go at the bottom? It’s more like an index; all of these fields should have been covered already by their respective sections. And the content for each field should just be a stub, maybe describing the field in a sentence but then linking to the relevant section.

* Files ending in `.cjs`.

* Files ending in `.js` when the nearest parent `package.json` file contains a
  top-level field [`"type"`][] with a value of `"commonjs"`.

* Strings passed in as an argument to `--eval` or `--print`, or piped to `node`
  via `STDIN`, with the flag `--input-type=commonjs`.

Package authors should include the [`"type"`][] field, even in packages where
all sources are CommonJS. Being explicit about the `type` of the package will
future-proof the package in case the default type of Node.js ever changes, and
it will also make things easier for build tools and loaders to determine how the
files in the package should be interpreted.

### Package scope and file extensions

A folder containing a `package.json` file, and all subfolders below that folder
until the next folder containing another [`package.json`][], are a
_package scope_. Package scopes do not carry through `node_modules` folders. The

> [`"type"`][] field defines how to treat `.js` files within the package scope. If
> a [`package.json`][] file does not have a [`"type"`][] field, the default is
> `"commonjs"`.

Reviewer 2: Perhaps something like:

```suggestion

Within the package scope, the [`"type"`][] field defines the module format for all `.js` files inside of that package scope.
The default for a scope with no type or an invalid type is `"commonjs"`.

Setting `"type": "module"` enables ES module syntax for all `.js` files inside of the scope.
```

I know we are repeating this info a lot, but it seems worth repeating.

* Files ending in `.cjs`.

* Files ending in `.js` when the nearest parent `package.json` file contains a
  top-level field [`"type"`][] with a value of `"commonjs"`.

* Strings passed in as an argument to `--eval` or `--print`, or piped to `node`
  via `STDIN`, with the flag `--input-type=commonjs`.

Package authors should include the [`"type"`][] field, even in packages where
all sources are CommonJS. Being explicit about the `type` of the package will
future-proof the package in case the default type of Node.js ever changes, and
it will also make things easier for build tools and loaders to determine how the
files in the package should be interpreted.

### Package scope and file extensions

A folder containing a `package.json` file, and all subfolders below that folder
until the next folder containing another [`package.json`][], are a
_package scope_. Package scopes do not carry through `node_modules` folders.

> Within the package scope, the [`"type"`][] field defines the module format for
> all `.js` files inside of that package scope. The default for a scope with no
> type or an invalid type is `"commonjs"`.

Author: ```suggestion
Within a package scope, the [`package.json`][] [`"type"`][] field defines how
Node.js should interpret `.js` files. If a `package.json` file does not have a
`"type"` field, `.js` files are treated as [CommonJS][].
```

}
```

## `package.json` supported fields

This document aims to describe the fields used by the Node.js
runtime. Other tools (such as
[npm](https://docs.npmjs.com/creating-a-package-json-file)) may use additional
fields which are ignored by Node.js and not documented here.

### `"name"`
<!-- YAML
added:
  - v12.16.0
  - v13.1.0
changes:
  - version:
    - v12.16.0
    - v13.6.0
    pr-url: https://github.com/nodejs/node/pull/31002
    description: Remove the `--experimental-resolve-self` option.
-->

```json
{
  "name": "package-name"
}
```

* Type: {string}

The `"name"` field defines your package’s name. Node.js doesn't apply any
restriction on the name field, although the field is ignored if it is not a

> string or an empty string.

Author: I've added the link to npm docs, PTAL

> string or an empty string.

Reviewer 2: I'd still prefer to remove this sentence -  "Node.js doesn't apply any
restriction on the name field, although the field is ignored if it is not a
string or an empty string."

> string or an empty string.

Author: OK, I personally feel we should have documented somewhere what are the limits Node.js enforces. Are you saying I should rephrase it or I should drop it because it's not relevant?

> string or an empty string.

Reviewer 2: Right, the thing is it is not quite this simple - because when loading self-resolution it must match the package name pattern of `/^(@[^@\/]+]\/)?[^@\/]+$/` and I'm not sure we should get into that here.

> string or an empty string.

Author: Interesting, I didn't know about the pattern! Goeffrey also agrees the sentence is irrelevant (https://github.com/nodejs/node/pull/34970#discussion_r483296692), I'm removing it.

where `import '#dep'` would now get the resolution of the external package
`dep-node-native` (including its exports in turn), and instead get the local
file `./dep-polyfill.js` relative to the package in other environments.

Unlike the exports field, import maps permit mapping to external packages
because this provides an important use case for conditional loading and also can
be done without the risk of cycles, unlike for exports.

Apart from the above, the resolution rules for the imports field are otherwise
analogous to the exports field.

### `"type"`
<!-- YAML
added: v12.0.0
changes:
  - version:
    - v12.17.0
    - v13.2.0
    pr-url: https://github.com/nodejs/node/pull/29866
    description: Unflag `--experimental-modules`.
-->

* Type: {string}

The `"type"` field defines how `.js` files should be treated
within a particular `package.json` file’s package scope.

Files ending with `.js` will be loaded as ES modules when the nearest parent
`package.json` file contains a top-level field `"type"` with a value of
`"module"`.

The nearest parent `package.json` is defined as the first `package.json` found
when searching in the current folder, that folder’s parent, and so on up

> until the root of the volume is reached.

Reviewer 2: ```suggestion
until a node_modules folder or the volume root is reached.
```

```json
// package.json
{
  "type": "module"
}
```

```console
# In same folder as above package.json
node my-app.js # Runs as ES module
```

> If the nearest parent `package.json` lacks a `"type"` field, or contains
> `"type": "commonjs"`, or any other unsupported value, `.js` files are treated as
> CommonJS. If the volume root is reached and no `package.json` is found, Node.js
> defers to the default, a `package.json` with no `"type"` field (CommonJS).

Reviewer 2: ```suggestion
If there is no package.json scope found, or if the nearest parent `package.json` lacks a `"type"` field, or contains
`"type": "commonjs"`, or any other unsupported value, `.js` files are treated as
CommonJS.
```

```js
// my-app.js, part of the same example as above
import './startup.js'; // Loaded as ES module because of package.json
```

Package authors should include the `"type"` field, even in packages where all
sources are CommonJS. Being explicit about the `type` of the package will
future-proof the package in case the default type of Node.js ever changes, and
it will also make things easier for build tools and loaders to determine how the

> files in the package should be interpreted.

Reviewer 2: This seems better to include back under scopes further up as this seems more "guideish".

I think the general rule should be facts here, advice up top?

Author: Moved to the Determining Module System section, PTAL

```

If there is no package.json scope found, or if the nearest parent `package.json`
lacks a `"type"` field, or contains `"type": "commonjs"`, or any other
unsupported value, `.js` files are treated as CommonJS.

```js
// my-app.js, part of the same example as above
import './startup.js'; // Loaded as ES module because of package.json
```

Regardless of the value of the `"type"` field, `.mjs` files are always treated
as ES modules and `.cjs` files are always treated as CommonJS.

### `"exports"`
<!-- YAML
added: v12.7.0
changes:
  - version:
    - v12.16.0
    - v13.2.0
    pr-url: https://github.com/nodejs/node/pull/29978
    description: Implement conditional exports.
  - version:
    - v12.15.0
    - v13.7.0
    pr-url: https://github.com/nodejs/node/pull/31001
    description: Remove the `--experimental-conditional-exports` option.
  - version:
    - v12.16.0
    - v13.7.0
    pr-url: https://github.com/nodejs/node/pull/31008
    description: Implement logical conditional exports ordering.
-->

```json
{
  "exports": "./index.js"
}
```

* Type: {Object} | {string} | {string[]}

The `"exports"` field allows defining the [entry points][] of a package when

> imported by name loaded either in a `node_modules` look or via a

Reviewer 2: ```suggestion
imported by name loaded either via a `node_modules` lookup or a
```

> [self-reference][] to its own name. It is supported in Node.js 12+ as an
> alternative to the [`"main"`][] that can support defining [subpath exports][]
> and [conditional exports][].

Reviewer 2: Sorry I left out the important encapsulation point here too!

```suggestion
[self-reference][] to its own name. It is supported in Node.js 12+ as an
alternative to the [`"main"`][] that can support defining [subpath exports][]
and [conditional exports][] while encapsulating internal unexported modules.
```

something like that? Open to suggestions.

Author: Oh I see, to get rid of the next paragraph. Yeah that's way more concise.

The `"name"` field defines your package’s name. Node.js doesn't apply any
restriction on the name field, although the field is ignored if it is not a
string or an empty string. Other tools (such as
[npm](https://docs.npmjs.com/files/package.json#name)) may apply other
restrictions on this field.

The `"name"` field can be used in addition to the [`"exports"`][] field to
[self-reference][] a package using its name.

### `"type"`
<!-- YAML
added: v12.0.0
changes:
  - version:
    - v12.17.0
    - v13.2.0
    pr-url: https://github.com/nodejs/node/pull/29866
    description: Unflag `--experimental-modules`.
-->

* Type: {string}

The `"type"` field defines the module format `.js` files should be treated
within a particular `package.json` file’s package scope.

Files ending with `.js` will be loaded as ES modules when the nearest parent
`package.json` file contains a top-level field `"type"` with a value of
`"module"`.

The nearest parent `package.json` is defined as the first `package.json` found
when searching in the current folder, that folder’s parent, and so on up
until a node_modules folder or the volume root is reached.

```json
// package.json
{
  "type": "module"
}
```

> ```console

Author: ```suggestion
```bash
```

> # In same folder as above package.json

Reviewer: ```suggestion
# In same folder as preceding package.json
```

There was a recent PR that changed references to "above" to "preceding" and similar. Please double-check that you're moving the latest versions of text. (This is just one example of several.)

node my-app.js # Runs as ES module
```

> If there is no package.json scope found, or if the nearest parent `package.json`
> lacks a `"type"` field, or contains `"type": "commonjs"`, or any other
> unsupported value, `.js` files are treated as CommonJS.

Reviewer: ```suggestion
If the nearest parent `package.json` lacks a `"type"` field, or contains	
`"type": "commonjs"`, `.js` files are treated as [CommonJS][]. If the volume	
root is reached and no `package.json` is found, Node.js defers to the default, a	
`package.json` with no `"type"` field.

`import` statements of `.js` files are treated as ES modules if the nearest	
parent `package.json` contains `"type": "module"`.
```

Let's keep the original text whenever possible please, it's gone through a lot of vetting already. "If there is no package.json scope found"? There's always a package scope. And let's leave out anything about unsupported values; there may be new types added in the future.

The second paragraph appears to have been lost in the reorganization, but it's important because some people assume that `import` always treats the target as ESM, but it doesn't.

Reviewer 2: This was at my suggestion to treat this section of the docs more like a comprehensive reference definition, while treating the top section of this page more like guide material. I actively encouraged refactoring as part of this process, just like refactoring when adding a feature I think it is natural and we shouldn't split out rewordings separately or risk not doing them at all.

Reviewer: > This was at my suggestion

Fair enough. I don't think this particular rewording is an improvement, though, and it covers the same content as the previous language.

Reviewer 2: The point was to make these definitions the comprehensive ones - as in to ensure concepts like the root scope are covered.

Reviewer 2: Then allow the top-of-the-page guide documentaiton to be able to skip these kinds of details rather.

Reviewer: > allow the top-of-the-page guide documentaiton to be able to skip these kinds of details rather.

This is in the bottom of the page . . . ? Are you opposed to my suggested change?

Reviewer 2: The only thing I would add is that I prefer to say the default treatment is commonjs than "no type" being a thing.

Otherwise the rewording seems fine to me.

```js
// my-app.js, part of the same example as above
import './startup.js'; // Loaded as ES module because of package.json
```

Regardless of the value of the `"type"` field, `.mjs` files are always treated
as ES modules and `.cjs` files are always treated as CommonJS.

> ### `"exports"`

Reviewer: Is this whole section copied from `esm.md` but not deleted from the old file? I don't think we want two copies of it.

Reviewer 2: We already moved all of esm over here. There is no exports section in esm anymore.

Reviewer: Then why is there no large block of deleted lines in this PR where the previous `"exports"` section came from? Currently it looks like this PR adds the entire `"exports"` section.

Reviewer 2: There is for "imports", but for "exports", this is replicating content from the upper guidance section. Again worth thinking about the reference v guide distinction. I still think reference here makes sense.

Author: There were no `"exports"` section in `esm.md`, the closest there was is the `Package entry point` section. This section is "new" content (or rather copied from #32970).

* Type: {Object} | {string} | {string[]}

The `"exports"` field allows defining the [entry points][] of a package when
imported by name loaded either via a `node_modules` lookup or a
[self-reference][] to its own name. It is supported in Node.js 12+ as an
alternative to the [`"main"`][] that can support defining [subpath exports][]
and [conditional exports][] while encapsulating internal unexported modules.

[Conditional Exports][] can also be used within `"exports"` to define different
package entry points per environment, including whether the package is
referenced via `require` or via `import`.

All paths defined in the `"exports"` must be relative file URLs starting with
`./`.

### `"main"`
<!-- YAML
added: v0.4.0
-->

```json
{
  "main": "./main.js"
}
```

* Type: {string}

The `"main"` field defines the script that is used when the [package directory
is loaded via `require()`](modules.html#modules_folders_as_modules). Its value
is interpreted as a path.

```js
require('./path/to/directory'); // This resolves to ./path/to/directory/main.js.
```

When a package has an [`"exports"`][] field, this will take precedence over the
`"main"` field when importing the package by name.

> ### `"imports"`

Reviewer: Is this whole section copied from `esm.md` but not deleted from the old file? I don't think we want two copies of it.

Reviewer 2: This was deleted from the top of this file, there is no imports section in esm.

      "node": "dep-node-native",
      "default": "./dep-polyfill.js"
    }
  },
  "dependencies": {
    "dep-node-native": "^1.0.0"
  }
}
```

where `import '#dep'` would now get the resolution of the external package
`dep-node-native` (including its exports in turn), and instead get the local
file `./dep-polyfill.js` relative to the package in other environments.

Unlike the exports field, import maps permit mapping to external packages
because this provides an important use case for conditional loading and also can
be done without the risk of cycles, unlike for exports.

Apart from the above, the resolution rules for the imports field are otherwise
analogous to the exports field.

[Conditional exports]: #packages_conditional_exports
[Babel]: https://babeljs.io/
[`esm`]: https://github.com/standard-things/esm#readme
[the full specifier path]: modules_esm.html#modules_esm_mandatory_file_extensions
[the dual CommonJS/ES module packages section]: #packages_dual_commonjs_es_module_packages
[ES modules]: esm.html
[ES module]: esm.html
[CommonJS]: modules.html
[`package.json`]: #packages_node_js_package_json_field_definitions
[`"exports"`]: #packages_exports
[`"main"`]: #packages_main

> [`"type"`]: #packages_type

Author: ```suggestion
[`"type"`]: #packages_type
[package scope]: #packages_package_scope_and_file_extensions
```

[self-reference]: #packages_self_referencing_a_package_using_its_name
[subpath exports]: #packages_subpath_exports

> [entry points]: #packages_package_entry_points

Author: There are actually no linter rules for that, but I agree it should be alphabetized nonetheless.

> [entry points]: #packages_package_entry_points

Reviewer: I think it’s part of CI maybe? I know I’ve gotten stung by this before.

[self-reference][] a package using its name.

### `"type"`
<!-- YAML
added: v12.0.0
changes:
  - version:
    - v12.17.0
    - v13.2.0
    pr-url: https://github.com/nodejs/node/pull/29866
    description: Unflag `--experimental-modules`.
-->

* Type: {string}

The `"type"` field defines the module format that Node.js will use for all
`.js` files within a particular `package.json` file’s [package scope][].

Files ending with `.js` will be loaded as ES modules when the nearest parent
`package.json` file contains a top-level field `"type"` with a value of
`"module"`.

The nearest parent `package.json` is defined as the first `package.json` found
when searching in the current folder, that folder’s parent, and so on up
until a node_modules folder or the volume root is reached.

```json
// package.json
{
  "type": "module"
}
```

```bash
# In same folder as preceding package.json
node my-app.js # Runs as ES module
```

If the nearest parent `package.json` lacks a `"type"` field, or contains	
`"type": "commonjs"`, `.js` files are treated as [CommonJS][]. If the volume	
root is reached and no `package.json` is found, Node.js defers to the default, a	

> `package.json` with no `"type"` field.

Author: ```suggestion
`package.json` with no `"type"` field and `.js` files are treated as [CommonJS][].
```

@guybedford would that work for you? If feels a bit repetitive but at least it's explicit.

Reviewer 2: Can we again just say "the default treatment as CommonJS" instead of explicitly a package.json with no type field?

> root is reached and no `package.json` is found, Node.js defers to the default, a	
> `package.json` with no `"type"` field.

Author: ```suggestion
root is reached and no `package.json` is found, Node.js defers to the default
treatment as [CommonJS][].
```
@guybedford is that what you mean?

Reviewer 2: Yes that sounds good thanks!
